File: src/main/java/de/visustruct/struktogrammelemente/ForSchleife.java
package de.visustruct.struktogrammelemente;

import java.awt.Graphics2D;

import de.visustruct.control.DiagramKeywordText;
import de.visustruct.control.GlobalSettings;
import de.visustruct.control.Struktogramm;
import de.visustruct.other.JTextAreaEasy;
import de.visustruct.view.CodeErzeuger;

public class ForSchleife extends WhileSchleife { //erbt von WhileSchleife
   private static final String KOPF_TRENNER = "; ";

   public ForSchleife(Graphics2D g){
      super(g);
      
      setzeText(GlobalSettings.gibElementBeschriftung(Struktogramm.typForSchleife));
   }

   @Override
   protected void randGroesseSetzen(){
      if (hatMehrteiligenKopf()){
         setObererRand(obererRandZusatz + gibTexthoehe(text[0]));
      }else{
         super.randGroesseSetzen();
      }
   }

   @Override
   protected int gibBreiteDerBreitestenTextzeile(){
      if (hatMehrteiligenKopf()){
         String display = DiagramKeywordText.lineForDisplay(Struktogramm.typForSchleife, 0, gibKopfText());
         if (objGesetzt(g)) {
            return DiagramKeywordText.measureLineWidth(g, display);
         }
         return Math.max(display.length() * 7, 4 * display.length());
      }

      return super.gibBreiteDerBreitestenTextzeile();
   }

   @Override
   protected void textZeichnen(){
      if (!hatMehrteiligenKopf()){
         super.textZeichnen();
         return;
      }

      String kopfText = gibKopfText();
      int texthoehe = gibTexthoehe(text[0]);

      String display = DiagramKeywordText.lineForDisplay(Struktogramm.typForSchleife, 0, kopfText);
      int x = gibX() + gibXVerschiebungFuerTextInMitte(0, display);
      DiagramKeywordText.drawKeywordAwareLine(g, getFarbeSchrift(), x, gibY() + texthoehe - 5, display);
   }

   private boolean hatMehrteiligenKopf(){
      return text != null && text.length > 1;
   }

   private String gibKopfText(){
      return String.join(KOPF_TRENNER, text);
   }
   
   
   @Override     //siehe DoUntilSchleife
   public void quellcodeGenerieren(int typ, int anzahlEingerueckt, int anzahlEinzuruecken, boolean alsKommentar, JTextAreaEasy textarea){
      String vorher = "";
      String nachher = "";


      if (typ == CodeErzeuger.typPython) {
         String pythonFor = CodeGenRules.pythonForLineFromJavaLikeLoopFields(text);
         if (pythonFor != null) {
            vorher = quellcodeMitKommentarVorspann("", "\n", pythonFor, typ, anzahlEingerueckt, alsKommentar);
         } else {
            vorher = quellcodeMitKommentarVorspann("for ", ":\n", typ, anzahlEingerueckt, alsKommentar);
         }
         nachher = "";
      } else if (typ == CodeErzeuger.typJavaScript && hatMehrteiligenKopf()) {
         vorher = quellcodeMitKommentarVorspann("for(", "){\n", CodeGenRules.javaScriptForLoopHeader(text), typ, anzahlEingerueckt, alsKommentar);
         nachher = "}\n";
      } else if (hatMehrteiligenKopf()) {
         vorher = quellcodeMitKommentarVorspann("for(", "){\n", CodeGenRules.cStyleForLoopHeader(text), typ, anzahlEingerueckt, alsKommentar);
         nachher = "}\n";
      } else {
         vorher = quellcodeMitKommentarVorspann("for(", "){\n", typ, anzahlEingerueckt, alsKommentar);
         nachher = "}\n";
      }

      textarea.hinzufuegen(wandleZuAusgabe(vorher,typ,anzahlEingerueckt,alsKommentar));
      liste.quellcodeAllerUnterelementeGenerieren(typ,anzahlEingerueckt+anzahlEinzuruecken,anzahlEinzuruecken,alsKommentar,textarea);
      if (!nachher.isEmpty()) {
         textarea.hinzufuegen(wandleZuAusgabe(nachher,typ,anzahlEingerueckt,alsKommentar));
      }

   }

   private String quellcodeMitKommentarVorspann(String linkerTeil, String rechterTeil, String codeText, int typ,
         int anzahlEingerueckt, boolean alsKommentar) {
      StringBuilder sb = new StringBuilder();
      if (alsKommentar) {
         sb.append(wandleZuAusgabe(co("kommentar") + co("text") + co("kommentarzu") + "\n", typ, anzahlEingerueckt, true));
      }
      sb.append(einruecken(linkerTeil + codeText + rechterTeil, anzahlEingerueckt));
      return sb.toString();
   }
   
}

File: src/main/kotlin/de/visustruct/struktogrammelemente/ForSchleife.kt
package de.visustruct.struktogrammelemente

import de.visustruct.control.DiagramKeywordText
import de.visustruct.control.GlobalSettings
import de.visustruct.control.Struktogramm
import de.visustruct.other.JTextAreaEasy
import de.visustruct.view.CodeErzeuger
import java.awt.Graphics2D

class ForSchleife(g: Graphics2D?) : WhileSchleife(g) {

    init {
        setzeText(GlobalSettings.gibElementBeschriftung(Struktogramm.typForSchleife))
    }

    override fun randGroesseSetzen() {
        if (hatMehrteiligenKopf()) {
            setObererRand(obererRandZusatz + gibTexthoehe(text[0]))
        } else {
            super.randGroesseSetzen()
        }
    }

    override fun gibBreiteDerBreitestenTextzeile(): Int {
        if (!hatMehrteiligenKopf()) {
            return super.gibBreiteDerBreitestenTextzeile()
        }
        val display = DiagramKeywordText.lineForDisplay(Struktogramm.typForSchleife, 0, gibKopfText())
        return if (objGesetzt(g)) {
            DiagramKeywordText.measureLineWidth(g, display)
        } else {
            maxOf(display.length * 7, 4 * display.length)
        }
    }

    override fun textZeichnen() {
        if (!hatMehrteiligenKopf()) {
            super.textZeichnen()
            return
        }
        val gfx = g ?: return
        val kopfText = gibKopfText()
        val texthoehe = gibTexthoehe(text[0])
        val display = DiagramKeywordText.lineForDisplay(Struktogramm.typForSchleife, 0, kopfText)
        val x = gibX() + gibXVerschiebungFuerTextInMitte(0, display)
        DiagramKeywordText.drawKeywordAwareLine(gfx, getFarbeSchrift(), x, gibY() + texthoehe - 5, display)
    }

    override fun quellcodeGenerieren(
        typ: Int,
        anzahlEingerueckt: Int,
        anzahlEinzuruecken: Int,
        alsKommentar: Boolean,
        textarea: JTextAreaEasy,
    ) {
        val vorher: String
        val nachher: String
        when {
            typ == CodeErzeuger.typPython -> {
                val pythonFor = CodeGenRules.pythonForLineFromJavaLikeLoopFields(text)
                vorher = if (pythonFor != null) {
                    quellcodeMitKommentarVorspann("", "\n", pythonFor, typ, anzahlEingerueckt, alsKommentar)
                } else {
                    quellcodeMitKommentarVorspann("for ", ":\n", typ, anzahlEingerueckt, alsKommentar)
                }
                nachher = ""
            }
            typ == CodeErzeuger.typJavaScript && hatMehrteiligenKopf() -> {
                vorher = quellcodeMitKommentarVorspann(
                    "for(", "){\n",
                    CodeGenRules.javaScriptForLoopHeader(text),
                    typ,
                    anzahlEingerueckt,
                    alsKommentar,
                )
                nachher = "}\n"
            }
            hatMehrteiligenKopf() -> {
                vorher = quellcodeMitKommentarVorspann(
                    "for(", "){\n",
                    CodeGenRules.cStyleForLoopHeader(text),
                    typ,
                    anzahlEingerueckt,
                    alsKommentar,
                )
                nachher = "}\n"
            }
            else -> {
                vorher = quellcodeMitKommentarVorspann("for(", "){\n", typ, anzahlEingerueckt, alsKommentar)
                nachher = "}\n"
            }
        }
        textarea.hinzufuegen(wandleZuAusgabe(vorher, typ, anzahlEingerueckt, alsKommentar))
        liste.quellcodeAllerUnterelementeGenerieren(
            typ,
            anzahlEingerueckt + anzahlEinzuruecken,
            anzahlEinzuruecken,
            alsKommentar,
            textarea,
        )
        if (nachher.isNotEmpty()) {
            textarea.hinzufuegen(wandleZuAusgabe(nachher, typ, anzahlEingerueckt, alsKommentar))
        }
    }

    private fun quellcodeMitKommentarVorspann(
        linkerTeil: String,
        rechterTeil: String,
        codeText: String,
        typ: Int,
        anzahlEingerueckt: Int,
        alsKommentar: Boolean,
    ): String {
        val sb = StringBuilder()
        if (alsKommentar) {
            sb.append(wandleZuAusgabe(co("kommentar") + co("text") + co("kommentarzu") + "\n", typ, anzahlEingerueckt, true))
        }
        sb.append(einruecken(linkerTeil + codeText + rechterTeil, anzahlEingerueckt))
        return sb.toString()
    }

    private fun hatMehrteiligenKopf(): Boolean = text.size > 1

    private fun gibKopfText(): String = text.joinToString(KOPF_TRENNER)

    companion object {
        private const val KOPF_TRENNER = "; "
    }
}
